lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,270 @@
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> void Post(const T &) {}
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    charBlock_ = statement.source;
+    if (statement.label.has_value()) {
+      labels_.insert(*statement.label);
+    }
+    return true;
+  }
+  // C1136
+  void Post(const parser::ReturnStmt &) {
+    messages_.Say(charBlock_,
+        parser::MessageFormattedText{
+            "RETURN not allowed in DO CONCURRENT"_err_en_US});
+  }
+  // C1137
+  void NoImageControl() {
+    messages_.Say(charBlock_,
+        parser::MessageFormattedText{
+            "image control statement not allowed in DO CONCURRENT"_err_en_US});
+  }
+  void Post(const parser::SyncAllStmt &) { NoImageControl(); }
+  void Post(const parser::SyncImagesStmt &) { NoImageControl(); }
+  void Post(const parser::SyncMemoryStmt &) { NoImageControl(); }
+  void Post(const parser::SyncTeamStmt &) { NoImageControl(); }
+  void Post(const parser::ChangeTeamConstruct &) { NoImageControl(); }
+  void Post(const parser::CriticalConstruct &) { NoImageControl(); }
+  void Post(const parser::EventPostStmt &) { NoImageControl(); }
+  void Post(const parser::EventWaitStmt &) { NoImageControl(); }
+  void Post(const parser::FormTeamStmt &) { NoImageControl(); }
+  void Post(const parser::LockStmt &) { NoImageControl(); }
+  void Post(const parser::UnlockStmt &) { NoImageControl(); }
+  void Post(const parser::StopStmt &) { NoImageControl(); }
+  void Post(const parser::EndProgramStmt &) { NoImageControl(); }
+
+  void Post(const parser::AllocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "ALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US});
+    }
+  }
+  void Post(const parser::DeallocateStmt &) {
+    if (anyObjectIsCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "DEALLOCATE coarray not allowed in DO CONCURRENT"_err_en_US});
+    }
+    // C1140: deallocation of polymorphic objects
+    if (anyObjectIsPolymorphic()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "DEALLOCATE polymorphic object(s)"
+              " not allowed in DO CONCURRENT"_err_en_US});
+    }
+  }
+  template<typename T> void Post(const parser::Statement<T> &) {
+    if (EndTDeallocatesCoarray()) {
+      messages_.Say(charBlock_,
+          parser::MessageFormattedText{
+              "implicit deallocation of coarray not allowed"
+              " in DO CONCURRENT"_err_en_US});
+    }
+  }
+  // C1141: cannot call ieee_get_flag, ieee_[gs]et_halting_mode
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name{std::get_if<parser::Name>(&procedureDesignator.u)}) {
+      // C1137: call move_alloc with coarray arguments
+      if (parser::ToUpperCaseLetters(name->ToString()) == "MOVE_ALLOC"s) {
```

Comment:
The case of identifiers is normalized, so `name->ToString() == "move_alloc"s` should be good enough.

Reply:
ok